### PR TITLE
Allow certain depackers to work without using a temporary file

### PR DESCRIPTION
**Pull request by ccawley2011**
This affects the ppdepack, s404, sqsh, muse, zip, gzip, arc and arcfs depackers. The rest still require a temporary file for now.

TODO:
 - [x] Fix the UBSan errors.
 - [x] Adapt the mmcmp depacker.
 - [x] Adapt the xz depacker.
 - [x] Adapt and test the xfd depacker.
 - [ ] Adapt the other depackers? (bzip2, compress, lha, lzx, rar, mo3)

### CODE REVIEW

File: src/hio.c
Comment (AliceLR) [inline src/hio.c:414]: It looks like `hio_open_mem` is used in a grand total of 2 places in load.c currently... might be able to get away with just making this take a `free_after_use` param like `mopen`.

**Comment by AliceLR** [comment]
> * Fix the UBSan errors.

I can look into these.

> * Adapt the mmcmp depacker.

I don't even get why this one needs a stream, the original implementation from libmodplug was entirely RAM...

> * Adapt the other depackers? (bzip2, compress, lha, lzx, xz, xfd, rar, mo3)

It might be better to hold off of adapting the others in this patch since they look generally more intrusive. RAR and MO3 are definitely each their own individual patches—RAR will probably need contrib code adapted from whatever has a permissively licensed unpacker (I *think* they exist, but The Unarchiver was LGPL...), MO3 iirc needs a proper UMX-style loader and MP3 decoder.

**Comment by AliceLR** [comment]
> * Fix the UBSan errors.

Ugh. in `tinfl_decompress_mem_to_heap` the values of `pBuf` and `pOut_len` are initially `NULL` and 0, and even include a ternary-guarded add, but `tinfl_decompress` then immediately adds them together without a guard. It looks like *maybe* the `tinfl_decompress` call should be skipped altogether if `pBuf` is NULL.

**Comment by sezero** [comment]
Is the issue known to mainstream miniz?

**Comment by AliceLR** [comment]
From a quick glance at the commit history and open issues/PRs, I don't see it.

**Comment by AliceLR** [comment]
The changes required to adapt the MMCMP depacker for this patch are merged now (#516).

**Comment by ccawley2011** [comment]
> The changes required to adapt the MMCMP depacker for this patch are merged now (#516).

Thanks, I've adapted this PR accordingly.

**Comment by sezero** [comment]
Reported the ubsan failure to mainstream at https://github.com/richgel999/miniz/issues/216

**Comment by sezero** [comment]
> Reported the ubsan failure to mainstream at [richgel999/miniz#216](https://github.com/richgel999/miniz/issues/216)

The following lazy patch seem to fix the issue - 
Also see: https://github.com/mitkus/miniz/commit/6b8c30e4a95dd649adf14cf4bcad90701e323b9a
```diff
diff --git a/src/depackers/miniz_tinfl.c b/src/depackers/miniz_tinfl.c
index 938ca09..11467be 100644
--- a/src/depackers/miniz_tinfl.c
+++ b/src/depackers/miniz_tinfl.c
@@ -186,7 +186,7 @@ tinfl_status tinfl_decompress(tinfl_decompressor *r, const mz_uint8 *pIn_buf_nex
     mz_uint32 num_bits, dist, counter, num_extra;
     tinfl_bit_buf_t bit_buf;
     const mz_uint8 *pIn_buf_cur = pIn_buf_next, *const pIn_buf_end = pIn_buf_next + *pIn_buf_size;
-    mz_uint8 *pOut_buf_cur = pOut_buf_next, *const pOut_buf_end = pOut_buf_next + *pOut_buf_size;
+    mz_uint8 *pOut_buf_cur = pOut_buf_next, *const pOut_buf_end = pOut_buf_next ? pOut_buf_next + *pOut_buf_size : NULL;
     size_t out_buf_size_mask = (decomp_flags & TINFL_FLAG_USING_NON_WRAPPING_OUTPUT_BUF) ? (size_t)-1 : ((pOut_buf_next - pOut_buf_start) + *pOut_buf_size) - 1, dist_from_out_buf_start;
 
     /* Ensure the output buffer's size is a power of 2, unless the output buffer is large enough to hold the entire output file (in which case it doesn't matter). */

```
Can you guys review?

**Comment by sezero** [comment]
I created a new [miniz2test branch](https://github.com/sezero/libxmp/tree/miniz2test) from #508, rebased to master, merged the patches from here, and applied the UB patch above: all tests pass. Patch submitted to mainstream as https://github.com/richgel999/miniz/pull/217

Off-topic to this PR: I also applied the mainstream [PR159 patchset](https://github.com/richgel999/miniz/pull/159) and it also passes all tests.

**Comment by ccawley2011** [comment]
I've updated the xz and xfd depackers following PR #522, however I'm not able to test any Amiga-specific changes. I've also cherry-picked the UBSan fix from the miniz2test branch, and CI now passes.

File: src/depackers/unxz.c
Comment (AliceLR) [inline src/depackers/unxz.c:95]: You should probably uncomment the above code to shrink the buffer first (I had a modified version of it that I didn't force push in time before it got merged):

```C
if ((tmp = (uint8 *) realloc(buf.out, buf.out_pos)) != NULL)
  buf.out = tmp;
```

**Comment by AliceLR** [comment]
The XFD and miniz updates seem OK to me. I don't have an Amiga toolchain set up so I can't test XFD either.

**Comment by sezero** [comment]
The xfd code fails build:
```
src/depackers/xfd.c: In function 'decrunch_xfd':
src/depackers/xfd.c:74: error: 'out' undeclared (first use in this function)
src/depackers/xfd.c:74: error: (Each undeclared identifier is reported only once
src/depackers/xfd.c:74: error: for each function it appears in.)
src/depackers/xfd.c: At top level:
src/depackers/xfd.c:99: warning: initialization from incompatible pointer type
make: *** [src/depackers/xfd.o] Error 1
```
(Note that the last warning is irrelevant.)

**Comment by sezero** [comment]
> (Note that the last warning is irrelevant.)

I take it back, it _is_ relevant.

**Comment by AliceLR** [comment]
More thorough look at the remaining depackers:
* unlha.c and unlzx.c are scary code and need to be replaced anyway. I don't think they should block this patch.
* bunzip2.c: might need an expanding buffer like unxz.c, I'm not familiar enough with the library. Also needs replacement, so it could probably be skipped for now.
* uncompress.c: It'd be nice if any of these ancient formats stored an uncompressed size, even an optional one. Should probably just implement a libxmp_vector.h at this rate or something to reduce the copypaste. ugh
* RAR: see above
* MO3: shouldn't be a depacker (see above)

**Comment by sezero** [comment]
XFD builds with the following patch:
```
--- a/src/depackers/xfd.c
+++ b/src/depackers/xfd.c
@@ -71,9 +71,9 @@ static int decrunch_xfd(HIO_HANDLE *f, v
 			if(xfdDecrunchBuffer(xfdobj))
 			{
 				unpacked = malloc(xfdobj->xfdbi_TargetBufSaveLen);
-				if (*out) {
+				if (*outbuf) {
 					memcpy(unpacked, xfdobj->xfdbi_TargetBuffer, xfdobj->xfdbi_TargetBufSaveLen);
-					*out = unpacked;
+					*outbuf = unpacked;
 					*outlen = xfdobj->xfdbi_TargetBufSaveLen;
 					ret=0;
 				}
@@ -96,8 +96,8 @@ static int decrunch_xfd(HIO_HANDLE *f, v
 
 struct depacker libxmp_depacker_xfd = {
 	test_xfd,
-	decrunch_xfd,
-	NULL
+	NULL,
+	decrunch_xfd
 };
 
 #endif /* AMIGA */
```

Build-tested only, and the patch constructed by quick comparison to
other changes, so needs a double check.

**Comment by sezero** [comment]
>     MO3: shouldn't be a depacker (see above)

Do we have mo3 code??

**Comment by AliceLR** [comment]
> EDIT: OK, that fork to external unmo3 tool, sorry..

unmo3 is a thing but it isn't maintained anymore (their README explicitly points to OpenMPT for updates now) and IMO it shouldn't be a depacker: what I've gathered looking into it is that it is its own module format rather than a container that happens to implement a common set of features for MOD/S3M/XM/IT, as well as structure compression, optional MP3 and Vorbis encoding, and a form of lossless encoding I haven't looked into much. Depending on the structure compression *maybe* it would partly work as a depacker, but all of the other aspects are much better suited to being a loader.

**Comment by ccawley2011** [comment]
The XFD changes should compile now.

**Comment by sezero** [comment]
> The XFD changes should compile now.

It does, thanks.

@AliceLR: What should we merge?

**Review by sezero — APPROVED** [review]

**Review by AliceLR — APPROVED** [review]
This looks OK to merge as-is. The remaining depackers are nontrivial and can be adapted later.

**Comment by sezero** [comment]
OK, merging. Will rebase and re-adjust #508 after it.

**Pull request merged**
